### PR TITLE
Spawn creature

### PR DESCRIPTION
**Pull request by bwrsandman**
image

extracting creature content out of https://github.com/openblack/openblack/pull/489

### CODE REVIEW

**Comment by raffclar** [comment]
@bwrsandman All good?

**Comment by raffclar** [comment]
I can squash the commits, up to you

File: src/3D/CreatureBody.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    {"c_ape", S::GiantApe}};
+
+using A = CreatureBody::Appearance;
+const std::map<std::string, CreatureBody::Appearance> k_MeshNameToAppearance = {
+    {"boned", A::Base}, {"base", A::Base},     {"good2", A::Good}, {"base2", A::Base}, {"good", A::Good}, {"evil", A::Evil},
+    {"evil2", A::Evil}, {"strong", A::Strong}, {"weak", A::Weak},  {"fat", A::Fat},    {"thin", A::Thin}};
+
+CreatureType creature::GetSpeciesFromMeshName(const std::string& name)
+{
+	for (const auto& [creatureTerm, species] : k_MeshNameToSpecies)
```
Comment (bwrsandman) [inline src/3D/CreatureBody.cpp:35]: Looping through a map to match a string?

File: src/3D/CreatureBody.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+			return species;
+		}
+	}
+
+	return CreatureType::Unknown;
+}
+
+CreatureBody::Appearance creature::GetAppearanceFromMeshName(const std::string& name)
+{
+	for (const auto& [creatureTerm, appearance] : k_MeshNameToAppearance)
```
Comment (bwrsandman) [inline src/3D/CreatureBody.cpp:48]: Looping through a map to match a string?

File: src/3D/CreatureBody.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			return appearance;
+		}
+	}
+
+	return CreatureBody::Appearance::Unknown;
+}
+
+entt::id_type creature::GetIdFromType(CreatureType species)
+{
+	auto defaultBodyType = CreatureBody::Appearance::Base;
```
Comment (bwrsandman) [inline src/3D/CreatureBody.cpp:61]: Any way to select another type?

File: src/ECS/Archetypes/CreatureArchetype.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+using namespace openblack::ecs::components;
+
+entt::entity CreatureArchetype::Create(const std::string& playerName, CreatureType creatureType,
+                                       const std::string& creatureMind, const glm::vec3& position)
+{
+	auto& registry = Game::Instance()->GetEntityRegistry();
+	const auto entity = registry.Create();
+	auto meshId = creature::GetIdFromType(creatureType);
+	registry.Assign<Creature>(entity, playerName, creatureType, creatureMind);
+	registry.Assign<ecs::components::Mesh>(entity, meshId);
```
Comment (bwrsandman) [inline src/ECS/Archetypes/CreatureArchetype.cpp:32]: Already `using namespace openblack::ecs::components`

File: src/ECS/Archetypes/CreatureArchetype.cpp [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	registry.Assign<Transform>(entity, position, glm::eulerAngleY(glm::radians(180.0f)), glm::vec3(.3f));
```
Comment (bwrsandman) [inline src/ECS/Archetypes/CreatureArchetype.cpp:33]: Rotation and scale are magic values? Where do they come from?

File: src/ECS/Components/Creature.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#include "3D/CreatureBody.h"
+#include "Creature.h"
+#include "Enums.h"
+
+namespace openblack::ecs::components
+{
+
+struct Creature
+{
+	std::string owner;
```
Comment (bwrsandman) [inline src/ECS/Components/Creature.h:23]: We have an enum for player name

File: src/ECS/Components/Creature.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	CreatureType species;
+	std::string mindName;
```
Comment (bwrsandman) [inline src/ECS/Components/Creature.h:25]: Is it possible to make an enum for this? Does this value map to anything in info.dat?

**Review by bwrsandman — COMMENTED** [review]
I have a few worries about performance here and you see to be misusing maps.

You can squash your "fixing this and that" commits, they don't useful information to the git history.

File: src/ECS/Archetypes/CreatureArchetype.cpp (hunk 5 above)
Comment (raffclar) [reply src/ECS/Archetypes/CreatureArchetype.cpp:33]: We use 180.0f elsewhere (StreetLanternArchetype, VillagerArchetype) to get meshes oriented correctly.
`.3f` gives the creature a reasonable size. 1.0f looks huge.

Comment (bwrsandman) [reply src/ECS/Archetypes/CreatureArchetype.cpp:33]: What does vanilla do?

File: src/3D/CreatureBody.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    {"evil2", A::Evil}, {"strong", A::Strong}, {"weak", A::Weak},  {"fat", A::Fat},    {"thin", A::Thin}};
+
+entt::id_type creature::GetIdFromType(CreatureType species)
+{
+	auto defaultBodyType = CreatureBody::Appearance::Base;
+	return entt::hashed_string(
+	    fmt::format("creature/{}/{}", static_cast<uint8_t>(species), static_cast<uint32_t>(defaultBodyType)).c_str());
+}
+
+entt::id_type creature::GetIdFromMeshName(const std::string& name)
```
Comment (bwrsandman) [inline src/3D/CreatureBody.cpp:42]: From reading this, it's not very clear what's going on.
Could you add a comment here explaining what the function is doing by splitting, searching, joining with `_` and then searching again.
Give an example.

File: src/Common/StringUtils.h [hunk 9]
```diff
@@ -25,4 +26,6 @@ namespace openblack::string_utils
 
 [[nodiscard]] std::string Capitalise(const std::string& str);
 
+[[nodiscard]] std::vector<std::string> Split(const std::string& string, const std::string& delimiter);
```
Comment (bwrsandman) [inline src/Common/StringUtils.h:29]: Could you look for more places where we split strings in the code and refactor them with this.

File: src/3D/CreatureBody.cpp (hunk 3 above)
Comment (raffclar) [reply src/3D/CreatureBody.cpp:37]: Not right now no.

File: src/ECS/Archetypes/CreatureArchetype.cpp (hunk 5 above)
Comment (raffclar) [reply src/ECS/Archetypes/CreatureArchetype.cpp:33]: Vanilla ignores the placement of the creatures and moves them to the temple. A good example of this is four gods where all A.I creatures are spawned at roughly the same location but are moved just afterwards to their pens.

image

Comment (bwrsandman) [reply src/ECS/Archetypes/CreatureArchetype.cpp:33]: And what scale do they start with?

File: src/3D/CreatureBody.cpp (hunk 3 above)
Comment (bwrsandman) [reply src/3D/CreatureBody.cpp:37]: Could you add it as a parameter?

File: src/Common/StringUtils.h (hunk 9 above)
Comment (raffclar) [reply src/Common/StringUtils.h:29]: Couldn't find anything. Any ideas?

**Comment by bwrsandman** [comment]
I recommend you rebase on top of temple-creature-citadel-base and work on merging #559 first

**Comment by bwrsandman** [comment]
I believe that the mesh flags will have a bit set on creatures that is not set on other meshes. I believe that it's probably a bit that says that there are eyes in the extra block. The same way that buildings have door position.
Would be nice to include eyes in this PR or a follow-up one

**Comment by raffclar** [comment]
> I believe that the mesh flags will have a bit set on creatures that is not set on other meshes. I believe that it's probably a bit that says that there are eyes in the extra block. The same way that buildings have door position. Would be nice to include eyes in this PR or a follow-up one

https://github.com/openblack/openblack/issues/575

File: src/ECS/Components/Creature.h (hunk 6 above)
Comment (raffclar) [reply src/ECS/Components/Creature.h]: Fixed.

File: src/3D/CreatureBody.cpp (hunk 1 above)
Comment (raffclar) [reply src/3D/CreatureBody.cpp]: Yeah... fixed

File: src/3D/CreatureBody.cpp (hunk 2 above)
Comment (raffclar) [reply src/3D/CreatureBody.cpp]: Same here. Fixed

File: src/ECS/Archetypes/CreatureArchetype.cpp (hunk 4 above)
Comment (raffclar) [reply src/ECS/Archetypes/CreatureArchetype.cpp]: Fixed.

File: src/3D/CreatureBody.cpp (hunk 3 above)
Comment (raffclar) [reply src/3D/CreatureBody.cpp:37]: Fixed.

File: src/ECS/Archetypes/CreatureArchetype.cpp (hunk 5 above)
Comment (raffclar) [reply src/ECS/Archetypes/CreatureArchetype.cpp:33]: The scale is copied from player one when playing against the A.I otherwise other players' creatures.

File: src/ECS/Components/Creature.h (hunk 7 above)
Comment (raffclar) [reply src/ECS/Components/Creature.h:25]: I really don't know enough about info.dat. Not inspected the file for creature mind stuff

File: src/3D/CreatureBody.cpp (hunk 8 above)
Comment (raffclar) [reply src/3D/CreatureBody.cpp:52]: I've reworked this slighty. I've also added docs to the header prototype.

**Comment by raffclar** [comment]
@bwrsandman Does this look ok?

File: src/ECS/Components/Creature.h (hunk 7 above)
Comment (bwrsandman) [reply src/ECS/Components/Creature.h:25]: What sort of values come out?

File: src/3D/CreatureBody.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include <map>
```
Comment (bwrsandman) [inline src/3D/CreatureBody.h:12]: Unnecessary big include in a header

File: src/3D/CreatureBody.h [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
```
Comment (bwrsandman) [inline src/3D/CreatureBody.h:13]: Unnecessary big include in a header
```suggestion
```

File: src/3D/CreatureBody.h [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+#include "Enums.h"
+
+namespace openblack
+{
+namespace ecs::components
+{
+struct Creature;
+}
+
```
Comment (bwrsandman) [inline src/3D/CreatureBody.h:24]: Unused forward declare
```suggestion
```

File: src/3D/CreatureBody.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include <map>
+
+#include <entt/entity/registry.hpp>
+
```
Comment (bwrsandman) [inline src/3D/CreatureBody.h:15]: Unnecessary big includes in a header
```suggestion
#include <entt/fwd.hpp>
```

File: src/ECS/Components/Creature.h [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include <string>
+
+#include "3D/CreatureBody.h"
```
Comment (bwrsandman) [inline src/ECS/Components/Creature.h:14]: Unused header
```suggestion
```

File: src/Game.cpp [hunk 15]
```diff
@@ -520,6 +521,27 @@ bool Game::Initialize()
...
 	}
 
+	const auto creatures = _fileSystem->FindPath(_fileSystem->CreatureMeshPath());
+	for (const auto& f : std::filesystem::directory_iterator {creatures})
+	{
+		const auto& fileName = f.path().stem().string();
+		SPDLOG_LOGGER_DEBUG(spdlog::get("game"), "Loading creature mesh: {}", fileName);
+		try
+		{
+			if (string_utils::BeginsWith(fileName, "Hand"))
```
[the remark this answers is not in the document]
Comment (raffclar) [reply src/Game.cpp:531]: Yeah. It's odd.

File: src/3D/CreatureBody.h (hunk 12 above)
Comment (raffclar) [reply src/3D/CreatureBody.h:24]: Well spotted

File: src/ECS/Components/Creature.h (hunk 7 above)
Comment (bwrsandman) [reply src/ECS/Components/Creature.h:25]: I see
* "ComputerControlledCreature"
* "COMPUTERCONTROLLEDCREATURE"

Comment (bwrsandman) [reply src/ECS/Components/Creature.h:25]: This seems to be for `Scripts\CreatureMind\ComputerControlledCreature`
The directory has other minds and I guess mods can add others
* ComputerControlledCreature
* KhazarCreature
* LethysCreature
* NemesisCreature

If I grep for `KhazarCreature`, I find references in my save files and in Scripts/Quests/challenge.chl

Comment (bwrsandman) [reply src/ECS/Components/Creature.h:25]: I've started decyphering erc files here:
https://github.com/bwrsandman/openblack/tree/erc

File: src/Common/stb_image_write.h [hunk 16]
```diff
@@ -314,7 +314,8 @@ static void stbiw__writefv(stbi__write_context* s, const char* fmt, va_list v)
 	{
```
Comment (bwrsandman) [inline src/Common/stb_image_write.h:314]: Why are these charges included?

File: src/Graphics/MetalViewHelper.mm [hunk 17]
```diff
@@ -7,17 +7,18 @@
  * openblack is licensed under the GNU General Public License version 3.
  *****************************************************************************/
 
-#import <QuartzCore/CAMetalLayer.h>
+#import <Cocoa/Cocoa.h>
```
Comment (bwrsandman) [inline src/Graphics/MetalViewHelper.mm:10]: Why are these changes included?

File: src/Creature/CreatureMind.h [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+ *****************************************************************************/
+
+#pragma once
+
+namespace openblack::creature
+{
+
+struct CreatureMind
+{
+};
```
Comment (bwrsandman) [inline src/Creature/CreatureMind.h:17]: Empty structs can cause problems with some compilers. I usually put a dummy uint8

File: src/ECS/Components/Creature.h [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+#include "Enums.h"
+
+namespace openblack::ecs::components
+{
+
+struct Creature
+{
+	PlayerNames owner;
+	CreatureType species;
+	entt::id_type creatureMindId;
```
Comment (bwrsandman) [inline src/ECS/Components/Creature.h:26]:
```suggestion
	entt::id_type mind;
```

**Comment by raffclar** [comment]
@bwrsandman ready for review
